File: apps/api/app/routers/projects.py
"""Project CRUD: list, get, create, update, soft-delete."""
import uuid
from typing import Annotated

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy import text
from sqlalchemy.ext.asyncio import AsyncSession

from app.auth import get_current_user_id
from app.database import get_db
from app.schemas.projects import ProjectCreate, ProjectResponse, ProjectUpdate

router = APIRouter()


def _row_to_project_response(row: dict) -> ProjectResponse:
    return ProjectResponse(
        id=row["id"],
        name=row["name"],
        description=row["description"],
        platform=row["platform"],
        head_version_id=row["head_version_id"],
        created_at=row["created_at"],
        updated_at=row["updated_at"],
    )


@router.get("", response_model=list[ProjectResponse])
async def list_projects(
    user_id: Annotated[uuid.UUID, Depends(get_current_user_id)],
    db: Annotated[AsyncSession, Depends(get_db)],
) -> list[ProjectResponse]:
    """List projects for the current user (exclude soft-deleted)."""
    result = await db.execute(
        text("""
            SELECT id, name, description, platform, head_version_id, created_at, updated_at
            FROM projects
            WHERE owner_id = :user_id AND deleted_at IS NULL
            ORDER BY updated_at DESC
        """),
        {"user_id": user_id},
    )
    rows = result.mappings().all()
    return [_row_to_project_response(dict(r)) for r in rows]


@router.get("/{project_id}", response_model=ProjectResponse)
async def get_project(
    project_id: uuid.UUID,
    user_id: Annotated[uuid.UUID, Depends(get_current_user_id)],
    db: Annotated[AsyncSession, Depends(get_db)],
) -> ProjectResponse:
    """Get one project; 404 if not found or not owner."""
    result = await db.execute(
        text("""
            SELECT id, name, description, platform, head_version_id, created_at, updated_at
            FROM projects
            WHERE id = :project_id AND owner_id = :user_id AND deleted_at IS NULL
        """),
        {"project_id": project_id, "user_id": user_id},
    )
    row = result.mappings().first()
    if not row:
        raise HTTPException(status_code=404, detail="Project not found")
    return _row_to_project_response(dict(row))


@router.post("", response_model=ProjectResponse, status_code=201)
async def create_project(
    body: ProjectCreate,
    user_id: Annotated[uuid.UUID, Depends(get_current_user_id)],
    db: Annotated[AsyncSession, Depends(get_db)],
) -> ProjectResponse:
    """Create a project; MVP sets platform to react-native."""
    result = await db.execute(
        text("""
            INSERT INTO projects (owner_id, name, description, platform, head_version_id, deleted_at, created_at, updated_at)
            VALUES (:owner_id, :name, :description, 'react-native', NULL, NULL, now(), now())
            RETURNING id, name, description, platform, head_version_id, created_at, updated_at
        """),
        {
            "owner_id": user_id,
            "name": body.name,
            "description": body.description,
        },
    )
    row = result.mappings().first()
    if not row:
        raise HTTPException(status_code=500, detail="Insert failed")
    await db.commit()
    return _row_to_project_response(dict(row))


@router.patch("/{project_id}", response_model=ProjectResponse)
async def update_project(
    project_id: uuid.UUID,
    body: ProjectUpdate,
    user_id: Annotated[uuid.UUID, Depends(get_current_user_id)],
    db: Annotated[AsyncSession, Depends(get_db)],
) -> ProjectResponse:
    """Update project name/description; 404 if not found or not owner."""
    # Verify ownership
    check = await db.execute(
        text("SELECT id FROM projects WHERE id = :id AND owner_id = :user_id AND deleted_at IS NULL"),
        {"id": project_id, "user_id": user_id},
    )
    if not check.mappings().first():
        raise HTTPException(status_code=404, detail="Project not found")
    # Build update: only non-None fields
    updates = []
    params: dict = {"project_id": project_id}
    if body.name is not None:
        updates.append("name = :name")
        params["name"] = body.name
    if body.description is not None:
        updates.append("description = :description")
        params["description"] = body.description
    if not updates:
        # No-op: fetch and return current
        result = await db.execute(
            text("""
                SELECT id, name, description, platform, head_version_id, created_at, updated_at
                FROM projects WHERE id = :project_id
            """),
            {"project_id": project_id},
        )
        row = result.mappings().first()
        if row:
            return _row_to_project_response(dict(row))
        raise HTTPException(status_code=404, detail="Project not found")
    updates.append("updated_at = now()")
    sql = f"UPDATE projects SET {', '.join(updates)} WHERE id = :project_id RETURNING id, name, description, platform, head_version_id, created_at, updated_at"
    result = await db.execute(text(sql), params)
    row = result.mappings().first()
    await db.commit()
    if not row:
        raise HTTPException(status_code=404, detail="Project not found")
    return _row_to_project_response(dict(row))


@router.delete("/{project_id}", status_code=204)
async def delete_project(
    project_id: uuid.UUID,
    user_id: Annotated[uuid.UUID, Depends(get_current_user_id)],
    db: Annotated[AsyncSession, Depends(get_db)],
) -> None:
    """Soft-delete project; 404 if not found or not owner."""
    result = await db.execute(
        text("""
            UPDATE projects
            SET deleted_at = now(), updated_at = now()
            WHERE id = :project_id AND owner_id = :user_id AND deleted_at IS NULL
        """),
        {"project_id": project_id, "user_id": user_id},
    )
    await db.commit()
    if result.rowcount == 0:
        raise HTTPException(status_code=404, detail="Project not found")

File: apps/web/src/lib/projects.ts
/**
 * Projects API client (M14).
 * Calls FastAPI project CRUD endpoints with Clerk token.
 */

import { getApiBaseUrl } from "./api";

export class ProjectApiError extends Error {
  constructor(
    message: string,
    public readonly status?: number
  ) {
    super(message);
    this.name = "ProjectApiError";
  }
}

export type GetToken = () => Promise<string | null>;

export interface ProjectResponse {
  id: string;
  name: string;
  description: string | null;
  platform: string;
  head_version_id: string | null;
  created_at: string;
  updated_at: string;
}

export interface ProjectCreate {
  name: string;
  description?: string | null;
}

export interface ProjectUpdate {
  name?: string | null;
  description?: string | null;
}

async function getAuthHeaders(getToken: GetToken): Promise<HeadersInit> {
  const token = await getToken();
  if (!token) {
    throw new ProjectApiError("Not signed in", 401);
  }
  return {
    Authorization: `Bearer ${token}`,
    "Content-Type": "application/json",
  };
}

async function handleResponse<T>(res: Response, parse: () => Promise<T>): Promise<T> {
  if (!res.ok) {
    let message = `Request failed (${res.status})`;
    try {
      const data = (await res.json()) as {
        detail?: string | string[];
        message?: string;
      };
      const detail = data.detail ?? data.message;
      if (typeof detail === "string") message = detail;
      else if (Array.isArray(detail)) message = detail.map(String).join("; ");
    } catch {
      // ignore parse error
    }
    throw new ProjectApiError(message, res.status);
  }
  return parse();
}

export async function listProjects(getToken: GetToken): Promise<ProjectResponse[]> {
  const base = getApiBaseUrl();
  const res = await fetch(`${base}/projects`, {
    method: "GET",
    headers: await getAuthHeaders(getToken),
  });
  return handleResponse(res, () => res.json());
}

export async function getProject(
  id: string,
  getToken: GetToken
): Promise<ProjectResponse> {
  const base = getApiBaseUrl();
  const res = await fetch(`${base}/projects/${id}`, {
    method: "GET",
    headers: await getAuthHeaders(getToken),
  });
  return handleResponse(res, () => res.json());
}

export async function createProject(
  body: ProjectCreate,
  getToken: GetToken
): Promise<ProjectResponse> {
  const base = getApiBaseUrl();
  const res = await fetch(`${base}/projects`, {
    method: "POST",
    headers: await getAuthHeaders(getToken),
    body: JSON.stringify(body),
  });
  return handleResponse(res, () => res.json());
}

export async function updateProject(
  id: string,
  body: ProjectUpdate,
  getToken: GetToken
): Promise<ProjectResponse> {
  const base = getApiBaseUrl();
  const res = await fetch(`${base}/projects/${id}`, {
    method: "PATCH",
    headers: await getAuthHeaders(getToken),
    body: JSON.stringify(body),
  });
  return handleResponse(res, () => res.json());
}

export async function deleteProject(
  id: string,
  getToken: GetToken
): Promise<void> {
  const base = getApiBaseUrl();
  const res = await fetch(`${base}/projects/${id}`, {
    method: "DELETE",
    headers: await getAuthHeaders(getToken),
  });
  if (!res.ok) {
    let message = `Request failed (${res.status})`;
    try {
      const data = (await res.json()) as { detail?: string | string[] };
      const detail = data.detail;
      if (typeof detail === "string") message = detail;
      else if (Array.isArray(detail)) message = detail.map(String).join("; ");
    } catch {
      // ignore
    }
    throw new ProjectApiError(message, res.status);
  }
}
